packages/next-swc/crates/napi/src/transform.rs

```diff
@@ -71,90 +71,99 @@ impl Task for TransformTask {
-                        color: ColorConfig::Always,
-                        skip_filename: skip_filename(),
-                    },
-                    |handler| {
-                        self.c.run(|| {
-                            let options: TransformOptions = serde_json::from_slice(&self.options)?;
-                            let fm = match &self.input {
-                                Input::Source { src } => {
-                                    let filename = if options.swc.filename.is_empty() {
-                                        FileName::Anon
-                                    } else {
-                                        FileName::Real(options.swc.filename.clone().into())
+        tokio::runtime::Runtime::new()
+            .unwrap()
+            .block_on(async move {
+                GLOBALS.set(&Default::default(), || {
+                    let eliminated_packages: Rc<RefCell<fxhash::FxHashSet<String>>> =
+                        Default::default();
+                    let res = catch_unwind(AssertUnwindSafe(|| {
+                        try_with_handler(
+                            self.c.cm.clone(),
+                            turbopack_binding::swc::core::base::HandlerOpts {
+                                color: ColorConfig::Always,
+                                skip_filename: skip_filename(),
+                            },
+                            |handler| {
+                                self.c.run(|| {
+                                    let options: TransformOptions =
+                                        serde_json::from_slice(&self.options)?;
+                                    let fm = match &self.input {
+                                        Input::Source { src } => {
+                                            let filename = if options.swc.filename.is_empty() {
+                                                FileName::Anon
+                                            } else {
+                                                FileName::Real(options.swc.filename.clone().into())
+                                            };
+
+                                            self.c.cm.new_source_file(filename, src.to_string())
+                                        }
+                                        Input::FromFilename => {
+                                            let filename = &options.swc.filename;
+                                            if filename.is_empty() {
+                                                bail!("no filename is provided via options");
+                                            }
+
+                                            self.c.cm.new_source_file(
+                                                FileName::Real(filename.into()),
+                                                read_to_string(filename).with_context(|| {
+                                                    format!(
```

Comment:
this was already way too much indentation before and now it's even worse :/

can you extract parts of it into a different function or something like that?